compiler/src/org.graalvm.compiler.core.aarch64/src/org/graalvm/compiler/core/aarch64/AArch64LIRGenerator.java

```diff
@@ -384,31 +385,44 @@ protected boolean emitCompare(PlatformKind cmpKind, Value a, Value b, Condition
                 bExt = arithmeticLIRGen.emitSignExtend(b, compareBytes * 8, 64);
             }
 
-            if (LIRValueUtil.isVariable(bExt)) {
-                left = load(bExt);
+            /*
+             * The AArch64 comparison instruction can treat register 31 as the stack pointer
+             * register for the left operand, but not for the right operand.
+             */
+            boolean aIsStackPointer = ValueUtil.isRegister(aExt) && ValueUtil.asRegister(aExt).equals(AArch64.sp);
+            boolean bIsStackPointer = ValueUtil.isRegister(bExt) && ValueUtil.asRegister(bExt).equals(AArch64.sp);
+
+            if (aIsStackPointer && bIsStackPointer) {
+                // move b to a non-sp register
```

Comment:
Ok, this turns "cmp(sp,sp)" into "mov(r10,sp); cmp(r10,sp)"
For cosmetic reasons, I want to say that you should update the right so that it does "mov(r10,sp); cmp(r10,r10)". But it's not going to make any difference.

Reply:
true. changed